src/core/thread/network_data_local.cpp

```diff
@@ -153,6 +153,68 @@ otError Local::RemoveHasRoutePrefix(const uint8_t *aPrefix, uint8_t aPrefixLengt
     return error;
 }
 
+otError Local::AddService(uint32_t aEnterpriseNumber, const uint8_t *aServiceData, uint8_t aServiceDataLength,
+                          bool aServerStable, const uint8_t *aServerData, uint8_t aServerDataLength)
+{
+    otError error = OT_ERROR_NONE;
+    ServiceTlv *serviceTlv;
+    ServerTlv *serverTlv;
+
+    RemoveService(aEnterpriseNumber, aServiceData, aServiceDataLength);
+
+    serviceTlv = reinterpret_cast<ServiceTlv *>(mTlvs + mLength);
+    Insert(reinterpret_cast<uint8_t *>(serviceTlv),
+           sizeof(ServiceTlv) + sizeof(uint8_t)/*mServiceDataLength*/
+           + ((aEnterpriseNumber == THREAD_ENTERPRISE_NUMBER) ? 0 : sizeof(uint32_t)) + aServiceDataLength
+           + aServerDataLength + sizeof(ServerTlv));
+
+    serviceTlv->Init();
+    serviceTlv->SetEnterpriseNumber(aEnterpriseNumber);
+    serviceTlv->SetServiceID(0);
+    serviceTlv->SetServiceData(aServiceData, aServiceDataLength);
+    serviceTlv->SetLength((sizeof(ServiceTlv) - sizeof(NetworkDataTlv)) + aServiceDataLength +
```

Comment:
Can we hide this arithmetic by using `ServiceTlv::SetServiceDataLength()` instead?

Reply:
I fail to see how could you hide the parts of this arithmetic, sadly. However, since `Insert()` call above had almost identical logic, I extracted common stuff to a variable and used it in both places.

Reply:
I guess implicitly part of the question is whether or not `ServiceTlv::SetServiceDataLength()` should also modify the TLV Length field.